File: dt_bidi_test.cpp
#include "nlink_utils.hpp"
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <cstring>
#include <unistd.h>
#include <fcntl.h>
#include <termios.h>
#include <poll.h>
#include <sys/ioctl.h>
#include <IOKit/serial/ioss.h>
#include <time.h>
#include <vector>

inline uint64_t nanotime() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (uint64_t)ts.tv_sec * 1000000000ULL + (uint64_t)ts.tv_nsec;
}

static int open_serial(const char* port, int baudrate) {
    int fd = open(port, O_RDWR | O_NOCTTY | O_NONBLOCK);
    if (fd < 0) { perror("open"); return -1; }
    struct termios tty;
    tcgetattr(fd, &tty);
    cfmakeraw(&tty);
    tty.c_cflag &= ~(CRTSCTS | PARENB | CSTOPB | CSIZE);
    tty.c_cflag |= CS8 | CREAD | CLOCAL;
    tty.c_iflag &= ~(IXON | IXOFF | IXANY);
    tty.c_oflag &= ~OPOST;
    tty.c_lflag &= ~(ICANON | ECHO | ECHOE | ISIG);
    tty.c_cc[VMIN] = 0;
    tty.c_cc[VTIME] = 0;
    cfsetspeed(&tty, B230400);
    tcsetattr(fd, TCSANOW, &tty);
    speed_t speed = (speed_t)baudrate;
    ioctl(fd, IOSSIOSPEED, &speed);
    int latency = 1;
    ioctl(fd, IOSSDATALAT, &latency);
    tcflush(fd, TCIOFLUSH);
    return fd;
}

struct Stats {
    uint64_t count = 0;
    double min = 1e9, max = 0, sum = 0, sum_sq = 0;
    void add(double v) {
        ++count; sum += v; sum_sq += v*v;
        if (v < min) min = v; if (v > max) max = v;
    }
    double avg() const { return count ? sum/count : 0; }
    double stddev() const {
        if (count < 2) return 0;
        double m = avg(), var = sum_sq/count - m*m;
        return var > 0 ? sqrt(var) : 0;
    }
};

// 从原始字节流中扫描 sync 头提取帧
static void scan_raw(uint8_t* buf, size_t& len, size_t payload_sz,
                     uint8_t sync0, uint8_t sync1,
                     std::vector<uint16_t>& seqs, std::vector<uint64_t>& recv_ns_list, uint64_t now_ns) {
    while (len >= payload_sz) {
        if (buf[0] == sync0 && buf[1] == sync1) {
            uint16_t seq = buf[2] | (buf[3] << 8);
            seqs.push_back(seq);
            recv_ns_list.push_back(now_ns);
            memmove(buf, buf + payload_sz, len - payload_sz);
            len -= payload_sz;
        } else {
            memmove(buf, buf + 1, len - 1);
            --len;
        }
    }
}

// 从原始流中提取 User_Frame1 帧
static void scan_user_frame1(uint8_t* buf, size_t& len,
                             std::vector<uint16_t>& seqs, std::vector<uint64_t>& recv_ns_list,
                             uint64_t now_ns, size_t payload_sz) {
    while (len >= 12) {  // min User_Frame1 size: 11 + 1 byte data
        if (buf[0] != 0x54 || buf[1] != 0xF1) {
            memmove(buf, buf + 1, len - 1); --len; continue;
        }
        if (len < 11) break;
        uint16_t dlen = buf[8] | (buf[9] << 8);
        size_t total = 11 + dlen;
        if (total > len) break;
        // verify checksum
        uint8_t cs = 0;
        for (size_t i = 0; i < total - 1; ++i) cs += buf[i];
        if (cs != buf[total - 1]) {
            memmove(buf, buf + 1, len - 1); --len; continue;
        }
        // extract payload
        if (dlen >= payload_sz && buf[10] == 0x5A && buf[11] == 0xA5) {
            uint16_t seq = buf[12] | (buf[13] << 8);
            seqs.push_back(seq);
            recv_ns_list.push_back(now_ns);
        }
        memmove(buf, buf + total, len - total);
        len -= total;
    }
}

int main(int argc, char** argv) {
    const char* mport = nullptr, *sport = nullptr;
    int slave_id = 0, baud = 921600, interval_ms = 10, count = 200, data_sz = 16;
    for (int i = 1; i < argc; ++i) {
        if (!strcmp(argv[i], "-m") && i+1<argc) mport = argv[++i];
        else if (!strcmp(argv[i], "-s") && i+1<argc) sport = argv[++i];
        else if (!strcmp(argv[i], "--id") && i+1<argc) slave_id = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-b") && i+1<argc) baud = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-i") && i+1<argc) interval_ms = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-c") && i+1<argc) count = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-d") && i+1<argc) data_sz = atoi(argv[++i]);
    }
    if (!mport || !sport) { fprintf(stderr, "用法: %s -m <master> -s <slave> [--id N] [-c N]\n", argv[0]); return 1; }
    if (data_sz < 12) data_sz = 12;

    fprintf(stderr, "DT_MODE0 双向延迟测试\n  MASTER: %s  SLAVE: %s  ID:%d  间隔:%dms  帧数:%d\n\n", mport, sport, slave_id, interval_ms, count);

    int mfd = open_serial(mport, baud);
    int sfd = open_serial(sport, baud);
    if (mfd < 0 || sfd < 0) return 1;

    // 建立 SLAVE→MASTER 双向联系
    {
        auto f = nlink::build_user_frame1(0x05, slave_id, nullptr, 0);
        write(mfd, f.data(), f.size());
        usleep(200000);
    }

    // 状态
    uint16_t seq = 0;
    uint64_t interval_ns = (uint64_t)interval_ms * 1000000ULL;
    uint64_t next_send_ns = nanotime() + interval_ns;
    bool done = false;

    struct Pending { uint64_t ms_send_ns; uint64_t sm_send_ns; };
    Pending pending[65536] = {};
    bool has_pending[65536] = {};

    Stats st_ms, st_sm, st_rt;
    uint64_t total_sent = 0, total_ms_rcv = 0, total_sm_rcv = 0, total_rt_done = 0;
    uint64_t lost_ms = 0, lost_sm = 0;

    // 接收缓冲
    uint8_t mbuf[65536], sbuf[65536];
    size_t mlen = 0, slen = 0;
    std::vector<uint8_t> payload(data_sz);

    struct pollfd pfds[2];
    pfds[0].fd = mfd; pfds[0].events = POLLIN;
    pfds[1].fd = sfd; pfds[1].events = POLLIN;

    uint64_t start_ns = nanotime();
    uint32_t display_seq = 0;

    fprintf(stdout, "%5s  %9s  %9s  %9s  %5s  %5s  %5s\n",
            "序号", "M→S(ms)", "S→M(ms)", "往返(ms)", "M→S丢", "S→M丢", "完成");
    fprintf(stdout, "------------------------------------------------------------\n");

    while (!done) {
        uint64_t now_ns = nanotime();
        int64_t wait_ns = (int64_t)(next_send_ns - now_ns);
        int timeout = (wait_ns > 0) ? (int)(wait_ns / 1000000) : 0;

        (void)poll(pfds, 2, timeout);
        now_ns = nanotime();

        // MASTER 接收 (S→M ACK)
        if (pfds[0].revents & POLLIN) {
            uint8_t rbuf[4096];
            ssize_t n = read(mfd, rbuf, sizeof(rbuf));
            if (n > 0 && mlen + n <= sizeof(mbuf)) {
                memcpy(mbuf + mlen, rbuf, n); mlen += n;
                std::vector<uint16_t> seqs; std::vector<uint64_t> ns_list;
                scan_user_frame1(mbuf, mlen, seqs, ns_list, now_ns, (size_t)data_sz);
                for (size_t i = 0; i < seqs.size(); ++i) {
                    uint16_t s = seqs[i];
                    if (has_pending[s]) {
                        double sm_ms = (double)(ns_list[i] - pending[s].sm_send_ns) / 1000000.0;
                        double rt_ms = (double)(ns_list[i] - pending[s].ms_send_ns) / 1000000.0;
                        st_sm.add(sm_ms); st_rt.add(rt_ms);
                        has_pending[s] = false;
                        ++total_sm_rcv; ++total_rt_done;
                    }
                }
            }
        }

        // SLAVE 接收 (M→S data)
        if (pfds[1].revents & POLLIN) {
            uint8_t rbuf[4096];
            ssize_t n = read(sfd, rbuf, sizeof(rbuf));
            if (n > 0 && slen + n <= sizeof(sbuf)) {
                memcpy(sbuf + slen, rbuf, n); slen += n;
                std::vector<uint16_t> seqs; std::vector<uint64_t> ns_list;
                scan_raw(sbuf, slen, (size_t)data_sz, 0xA5, 0x5A, seqs, ns_list, now_ns);
                for (size_t i = 0; i < seqs.size(); ++i) {
                    uint16_t s = seqs[i];
                    if (has_pending[s]) {
                        double ms_ms = (double)(ns_list[i] - pending[s].ms_send_ns) / 1000000.0;
                        st_ms.add(ms_ms);
                        ++total_ms_rcv;

                        // 发送 ACK: S→M
                        uint8_t ack_payload[16];
                        memset(ack_payload, 0, data_sz);
                        ack_payload[0] = 0x5A; ack_payload[1] = 0xA5;
                        ack_payload[2] = s & 0xFF; ack_payload[3] = (s >> 8) & 0xFF;
                        auto ack = nlink::build_user_frame1(0x05, slave_id, ack_payload, data_sz);
                        pending[s].sm_send_ns = nanotime();
                        write(sfd, ack.data(), ack.size());
                    }
                }
            }
        }

        // 超时清理 M→S
        {
            uint64_t deadline = now_ns - 5000000000ULL; // 5s
            for (int i = 0; i < 65536; ++i)
                if (has_pending[i] && pending[i].ms_send_ns < deadline)
                    { has_pending[i] = false; ++lost_ms; }
        }

        // 发送 M→S
        if (now_ns >= next_send_ns && !done) {
            if (count == 0 || seq < (uint16_t)count) {
                nlink::pack_payload(payload.data(), data_sz, seq, (double)nanotime()/1e9);
                auto frame = nlink::build_unicast_frame(slave_id, payload.data(), data_sz);
                pending[seq].ms_send_ns = nanotime();
                write(mfd, frame.data(), frame.size());
                has_pending[seq] = true;
                ++total_sent;

                if (seq > display_seq) {
                    display_seq = seq;
                    fprintf(stdout, "%5u  %9.3f  %9.3f  %9.3f  %5llu  %5llu  %5llu\n",
                            seq, st_ms.avg(), st_sm.avg(), st_rt.avg(),
                            (unsigned long long)lost_ms, (unsigned long long)lost_sm,
                            (unsigned long long)total_rt_done);
                    fflush(stdout);
                }

                ++seq;
                next_send_ns += interval_ns;
                if (next_send_ns < now_ns) next_send_ns = now_ns + interval_ns;
            }
            if (count > 0 && seq >= (uint16_t)count) {
                bool any = false;
                for (int i = 0; i < 65536; ++i) if (has_pending[i]) { any = true; break; }
                if (!any) done = true;
                else if (now_ns - start_ns > (uint64_t)count * interval_ns + 10000000000ULL) done = true;
            }
        }
    }

    close(mfd); close(sfd);
    double total_s = (double)(nanotime() - start_ns) / 1e9;

    fprintf(stdout, "\n==============================\n");
    fprintf(stdout, "       双向延迟测试报告\n");
    fprintf(stdout, "==============================\n");
    fprintf(stdout, "  发送: %llu  M→S收到: %llu  S→M收到: %llu  往返完成: %llu\n",
            (unsigned long long)total_sent, (unsigned long long)total_ms_rcv,
            (unsigned long long)total_sm_rcv, (unsigned long long)total_rt_done);
    fprintf(stdout, "  总耗时: %.1fs\n", total_s);
    fprintf(stdout, "  --------------------------------\n");
    if (st_ms.count) {
        fprintf(stdout, "  M→S  平均: %8.3f ms  最小: %8.3f  最大: %8.3f  σ: %.3f\n",
                st_ms.avg(), st_ms.min, st_ms.max, st_ms.stddev());
    }
    if (st_sm.count) {
        fprintf(stdout, "  S→M  平均: %8.3f ms  最小: %8.3f  最大: %8.3f  σ: %.3f\n",
                st_sm.avg(), st_sm.min, st_sm.max, st_sm.stddev());
    }
    if (st_rt.count) {
        fprintf(stdout, "  往返 平均: %8.3f ms  最小: %8.3f  最大: %8.3f  σ: %.3f\n",
                st_rt.avg(), st_rt.min, st_rt.max, st_rt.stddev());
    }
    fprintf(stdout, "==============================\n");
    return 0;
}

File: rust/src/dt_bidi_test.rs
use clap::Parser;
use serialport::SerialPort;
use std::io::{Read, Write};
use std::time::Instant;
use uwb_delay_test::{nlink, pack_payload, Stats, SYNC0, SYNC1, PAYLOAD_HEADER_SIZE};

#[derive(Parser, Debug)]
#[command(name = "dt_bidi_test")]
struct Args {
    #[arg(short = 'm', long)]
    master_port: String,

    #[arg(short = 's', long)]
    slave_port: String,

    #[arg(long, default_value = "0")]
    id: u8,

    #[arg(short = 'b', long, default_value = "921600")]
    baudrate: u32,

    #[arg(short = 'i', long, default_value = "10")]
    interval_ms: u64,

    #[arg(short = 'c', long, default_value = "200")]
    count: u16,

    #[arg(short = 'd', long, default_value = "16")]
    data_size: usize,
}

fn open_serial(port: &str, baudrate: u32) -> Box<dyn SerialPort> {
    serialport::new(port, baudrate)
        .timeout(std::time::Duration::from_secs(0))
        .open()
        .unwrap_or_else(|e| panic!("cannot open {}: {}", port, e))
}

#[derive(Clone, Copy)]
struct Pending {
    ms_send_ns: u64,
    sm_send_ns: u64,
    active: bool,
}

fn main() {
    let args = Args::parse();

    let data_size = if args.data_size < PAYLOAD_HEADER_SIZE { PAYLOAD_HEADER_SIZE } else { args.data_size };

    eprintln!("DT_MODE0 bidirectional delay test (Rust optimized)");
    eprintln!("  MASTER: {}  SLAVE: {}  ID:{}  interval:{}ms  count:{}",
        args.master_port, args.slave_port, args.id, args.interval_ms, args.count);
    eprintln!();

    let mut master = open_serial(&args.master_port, args.baudrate);
    let mut slave = open_serial(&args.slave_port, args.baudrate);

    {
        let f = nlink::build_user_frame1(nlink::Role::Slave as u8, args.id, &[]);
        let _ = master.write_all(&f);
        std::thread::sleep(std::time::Duration::from_millis(200));
    }

    let mut seq: u16 = 0;
    let interval_ns = args.interval_ms * 1_000_000;
    let start_instant = Instant::now();
    let mut next_send_ns: u64 = 0;
    let mut done = false;

    let mut pending: Box<[Pending; 65536]> = Box::new([Pending { ms_send_ns: 0, sm_send_ns: 0, active: false }; 65536]);

    let mut st_ms = Stats::new();
    let mut st_sm = Stats::new();
    let mut st_rt = Stats::new();

    let mut total_sent: u64 = 0;
    let mut total_ms_rcv: u64 = 0;
    let mut total_sm_rcv: u64 = 0;
    let mut total_rt_done: u64 = 0;
    let mut lost_ms: u64 = 0;

    let mut mbuf: Box<[u8; 65536]> = Box::new([0u8; 65536]);
    let mut mlen: usize = 0;
    let mut sbuf: Box<[u8; 65536]> = Box::new([0u8; 65536]);
    let mut slen: usize = 0;
    
    let mut payload = vec![0u8; data_size];
    let mut ack_payload = vec![0u8; data_size];
    ack_payload[0] = 0x5A;
    ack_payload[1] = 0xA5;

    let mut display_seq: u32 = 0;

    println!("{:>5}  {:>9}  {:>9}  {:>9}  {:>5}  {:>5}  {:>5}",
        "seq", "M-S(ms)", "S-M(ms)", "RT(ms)", "M-S-lost", "S-M-lost", "done");
    println!("{}", "-".repeat(60));

    let mut read_buf = [0u8; 4096];
    let mut active_seqs: Vec<u16> = Vec::with_capacity(256);

    while !done {
        let now_ns = start_instant.elapsed().as_nanos() as u64;

        if let Ok(n) = master.read(&mut read_buf) {
            if n > 0 && mlen + n <= mbuf.len() {
                mbuf[mlen..mlen + n].copy_from_slice(&read_buf[..n]);
                mlen += n;

                for result in scan_user_frame1(&mut mbuf, &mut mlen, data_size, now_ns) {
                    let (s, recv_ns) = result;
                    if pending[s as usize].active {
                        let sm_ms = (recv_ns - pending[s as usize].sm_send_ns) as f64 / 1_000_000.0;
                        let rt_ms = (recv_ns - pending[s as usize].ms_send_ns) as f64 / 1_000_000.0;
                        st_sm.add(sm_ms);
                        st_rt.add(rt_ms);
                        pending[s as usize].active = false;
                        total_sm_rcv += 1;
                        total_rt_done += 1;
                    }
                }
            }
        }

        if let Ok(n) = slave.read(&mut read_buf) {
            if n > 0 && slen + n <= sbuf.len() {
                sbuf[slen..slen + n].copy_from_slice(&read_buf[..n]);
                slen += n;

                for result in scan_raw(&mut sbuf, &mut slen, data_size, now_ns) {
                    let (s, recv_ns) = result;
                    if pending[s as usize].active {
                        let ms_ms = (recv_ns - pending[s as usize].ms_send_ns) as f64 / 1_000_000.0;
                        st_ms.add(ms_ms);
                        total_ms_rcv += 1;

                        ack_payload[2] = (s & 0xFF) as u8;
                        ack_payload[3] = ((s >> 8) & 0xFF) as u8;
                        let ack = nlink::build_user_frame1(nlink::Role::Slave as u8, args.id, &ack_payload);
                        pending[s as usize].sm_send_ns = start_instant.elapsed().as_nanos() as u64;
                        let _ = slave.write_all(&ack);
                    }
                }
            }
        }

        if !active_seqs.is_empty() {
            let deadline = now_ns.saturating_sub(5_000_000_000);
            active_seqs.retain(|&s| {
                if pending[s as usize].ms_send_ns < deadline {
                    pending[s as usize].active = false;
                    lost_ms += 1;
                    false
                } else {
                    true
                }
            });
        }

        if now_ns >= next_send_ns && !done {
            if args.count == 0 || seq < args.count {
                pack_payload(&mut payload, seq, now_ns as f64 / 1e9);
                let frame = nlink::build_unicast_frame(args.id, &payload);
                pending[seq as usize].ms_send_ns = start_instant.elapsed().as_nanos() as u64;
                let _ = master.write_all(&frame);
                pending[seq as usize].active = true;
                active_seqs.push(seq);
                total_sent += 1;

                if seq as u32 > display_seq {
                    display_seq = seq as u32;
                    println!("{:>5}  {:>9.3}  {:>9.3}  {:>9.3}  {:>5}  {:>5}  {:>5}",
                        seq, st_ms.avg_ms(), st_sm.avg_ms(), st_rt.avg_ms(),
                        lost_ms, 0, total_rt_done);
                }

                seq += 1;
                next_send_ns += interval_ns;
                if next_send_ns < now_ns {
                    next_send_ns = now_ns + interval_ns;
                }
            }

            if args.count > 0 && seq >= args.count {
                if active_seqs.is_empty() {
                    done = true;
                } else if now_ns > (args.count as u64) * interval_ns + 10_000_000_000 {
                    done = true;
                }
            }
        }
    }

    let total_s = start_instant.elapsed().as_secs_f64();

    println!();
    println!("{}", "=".repeat(30));
    println!("   BIDIRECTIONAL TEST REPORT");
    println!("{}", "=".repeat(30));
    println!("  sent: {}  M-S-recv: {}  S-M-recv: {}  RT-done: {}",
        total_sent, total_ms_rcv, total_sm_rcv, total_rt_done);
    println!("  total time: {:.1}s", total_s);
    println!("  --------------------------------");
    if st_ms.count > 0 {
        println!("  M-S  avg: {:8.3} ms  min: {:8.3}  max: {:8.3}  sigma: {:.3}",
            st_ms.avg_ms(), st_ms.min_ms, st_ms.max_ms, st_ms.stddev_ms());
    }
    if st_sm.count > 0 {
        println!("  S-M  avg: {:8.3} ms  min: {:8.3}  max: {:8.3}  sigma: {:.3}",
            st_sm.avg_ms(), st_sm.min_ms, st_sm.max_ms, st_sm.stddev_ms());
    }
    if st_rt.count > 0 {
        println!("  RT   avg: {:8.3} ms  min: {:8.3}  max: {:8.3}  sigma: {:.3}",
            st_rt.avg_ms(), st_rt.min_ms, st_rt.max_ms, st_rt.stddev_ms());
    }
    println!("{}", "=".repeat(30));
}

struct ScanIter<'a> {
    buf: &'a mut [u8; 65536],
    len: &'a mut usize,
    payload_sz: usize,
    now_ns: u64,
}

impl<'a> Iterator for ScanIter<'a> {
    type Item = (u16, u64);

    fn next(&mut self) -> Option<Self::Item> {
        while *self.len >= self.payload_sz {
            if self.buf[0] == SYNC0 && self.buf[1] == SYNC1 {
                let seq = self.buf[2] as u16 | ((self.buf[3] as u16) << 8);
                self.buf.copy_within(self.payload_sz..*self.len, 0);
                *self.len -= self.payload_sz;
                return Some((seq, self.now_ns));
            } else {
                self.buf.copy_within(1..*self.len, 0);
                *self.len -= 1;
            }
        }
        None
    }
}

#[inline]
fn scan_raw<'a>(buf: &'a mut [u8; 65536], len: &'a mut usize, payload_sz: usize, now_ns: u64) -> ScanIter<'a> {
    ScanIter { buf, len, payload_sz, now_ns }
}

struct UserFrameIter<'a> {
    buf: &'a mut [u8; 65536],
    len: &'a mut usize,
    payload_sz: usize,
    now_ns: u64,
}

impl<'a> Iterator for UserFrameIter<'a> {
    type Item = (u16, u64);

    fn next(&mut self) -> Option<Self::Item> {
        while *self.len >= 12 {
            if self.buf[0] != 0x54 || self.buf[1] != 0xF1 {
                self.buf.copy_within(1..*self.len, 0);
                *self.len -= 1;
                continue;
            }
            if *self.len < 11 {
                break;
            }
            let dlen = self.buf[8] as usize | ((self.buf[9] as usize) << 8);
            let total = 11 + dlen;
            if total > *self.len {
                break;
            }

            let cs = self.buf[..total - 1].iter().fold(0u8, |a, &b| a.wrapping_add(b));
            if cs != self.buf[total - 1] {
                self.buf.copy_within(1..*self.len, 0);
                *self.len -= 1;
                continue;
            }

            let result = if dlen >= self.payload_sz && self.buf[10] == 0x5A && self.buf[11] == 0xA5 {
                let seq = self.buf[12] as u16 | ((self.buf[13] as u16) << 8);
                Some((seq, self.now_ns))
            } else {
                None
            };
            
            self.buf.copy_within(total..*self.len, 0);
            *self.len -= total;
            
            if result.is_some() {
                return result;
            }
        }
        None
    }
}

#[inline]
fn scan_user_frame1<'a>(buf: &'a mut [u8; 65536], len: &'a mut usize, payload_sz: usize, now_ns: u64) -> UserFrameIter<'a> {
    UserFrameIter { buf, len, payload_sz, now_ns }
}
